### PR TITLE
Update Python docs

### PR DESCRIPTION
**Pull request by creemama**
Note that after this pull request, I'll create another one for PHP cleanup: https://github.com/creemama/libopaque/commits/php.

### CODE REVIEW

File: python/test/simple.py [hunk 1]
```diff
@@ -24,44 +24,44 @@
 # password can be checked on the server for password rules
 # (e.g. occurence in common password lists), has the drawback that the
 # password is exposed to the server.
-rec, export_key0 = opaque.Register(pwd,cfg,ids,skS=None)
+rec, export_key = opaque.Register(pwdU,cfg,ids,skS=None)
 
 # user initiates a credential request
-pub, sec = opaque.CreateCredentialRequest(pwd)
+pub, sec0 = opaque.CreateCredentialRequest(pwdU)
```
Comment (stef) [inline python/test/simple.py:33]: secU might be better here indicating ownership

File: python/test/simple.py [hunk 2]
```diff
@@ -24,44 +24,44 @@
...
 
 # server responds to credential request
-resp, sks, ctx = opaque.CreateCredentialResponse(pub, rec, cfg, ids, None)
+resp, sk, sec1 = opaque.CreateCredentialResponse(pub, rec, cfg, ids, None)
```
Comment (stef) [inline python/test/simple.py:36]: this would be secS then

File: python/test/simple.py [hunk 3]
```diff
@@ -24,44 +24,44 @@
 # password can be checked on the server for password rules
 # (e.g. occurence in common password lists), has the drawback that the
 # password is exposed to the server.
-rec, export_key0 = opaque.Register(pwd,cfg,ids,skS=None)
+rec, export_key = opaque.Register(pwdU,cfg,ids,skS=None)
```
Comment (stef) [inline python/test/simple.py:30]: by calling it export_key0 and later having also an export_key, we can do an assert(==)

File: python/test/simple.py [hunk 4]
```diff
@@ -24,44 +24,44 @@
...
-pub, sec = opaque.CreateCredentialRequest(pwd)
+pub, sec0 = opaque.CreateCredentialRequest(pwdU)
 
 # server responds to credential request
-resp, sks, ctx = opaque.CreateCredentialResponse(pub, rec, cfg, ids, None)
+resp, sk, sec1 = opaque.CreateCredentialResponse(pub, rec, cfg, ids, None)
 
 # user recovers its credentials from the servers response
-sku, auth, export_key1, ids = opaque.RecoverCredentials(resp, sec, cfg, None, pkS=None)
+sk, authU, export_key, ids = opaque.RecoverCredentials(resp, sec0, cfg, None, pkS=None)
```
Comment (stef) [inline python/test/simple.py:39]: i guess there is an assert missing here checking that ek0 == ek

File: python/test/simple.py [hunk 5]
```diff
@@ -24,44 +24,44 @@
...
-ctx, alpha = opaque.CreateRegistrationRequest(pwd)
+sec0, M = opaque.CreateRegistrationRequest(pwdU)
 
 # server responds to the registration request
-sec, pub = opaque.CreateRegistrationResponse(alpha)
+sec1, pub = opaque.CreateRegistrationResponse(M)
 
 # user finalizes the registration using the response from the server
-rec, export_key3 = opaque.FinalizeRequest(ctx, pub, cfg, ids)
+rec, export_key = opaque.FinalizeRequest(sec0, pub, cfg, ids)
```
Comment (stef) [inline python/test/simple.py:53]: again we should assert that ek3 == ek0

File: python/test/simple.py [hunk 6]
```diff
@@ -24,44 +24,44 @@
...
-pub, sec = opaque.CreateCredentialRequest(pwd)
+pub, sec0 = opaque.CreateCredentialRequest(pwdU)
 
 # 2. server responds to credential request
-resp, sks, ctx = opaque.CreateCredentialResponse(pub, rec, cfg, ids, None)
+resp, sk, sec1 = opaque.CreateCredentialResponse(pub, rec, cfg, ids, None)
 
 # 3. user recovers its credentials from the server ressponse
-sku, auth, export_key4, ids = opaque.RecoverCredentials(resp, sec, cfg, None, pkS=None)
+sk, authU, export_key, ids = opaque.RecoverCredentials(resp, sec0, cfg, None, pkS=None)
```
Comment (stef) [inline python/test/simple.py:65]: another assert missing here

File: python/opaque/__init__.py [hunk 7]
```diff
@@ -300,14 +300,34 @@ def RecoverCredentials(resp, sec, cfg, infos, pkS=None):
...
+        if not ids.idS:
+            raise ValueError("ids.idU cannot be None if cfg.idS is NotPackaged.")
+        ids1.idS=ids.idS
+        ids1.idS_len=ids.idS_len
+    else:
+        ids1.idS=ctypes.cast(ctypes.create_string_buffer(1024), ctypes.c_char_p)
+        ids1.idS_len=1024
+
+    __check(opaquelib.opaque_RecoverCredentials(resp, sec, pkS, ctypes.pointer(cfg), ctypes.pointer(infos) if infos else None, ctypes.pointer(ids1), sk, authU, export_key))
+    return sk.raw, authU.raw, export_key.raw, ids1
```
Comment (stef) [inline python/opaque/__init__.py]: this is a nice one!

File: python/opaque/__init__.py [hunk 8]
```diff
@@ -414,7 +434,7 @@ def Create1kRegistrationResponse(M, pkS):
     if None in  (M, pkS):
         raise ValueError("invalid parameter")
     if len(M) != 32: raise ValueError("invalid M param")
-    if len(pkS) != pkS: raise ValueError("invalid pkS param")
+    if len(pkS) != crypto_scalarmult_BYTES: raise ValueError("invalid pkS param")
```
Comment (stef) [inline python/opaque/__init__.py]: wtf... /o\

File: python/opaque/__init__.py [hunk 9]
```diff
@@ -527,7 +547,7 @@ def Store1kUserRecord(sec, skS, rec):
     if len(skS) != crypto_scalarmult_SCALARBYTES: raise ValueError("invalid skS param")
     if len(rec) <= OPAQUE_USER_RECORD_LEN: raise ValueError("invalid rec param")
 
-    opaquelib.opaque_StoreUserRecord(sec, skS, rec)
+    opaquelib.opaque_Store1kUserRecord(sec, skS, rec)
```
Comment (stef) [inline python/opaque/__init__.py]: hrmpf. /o\
thanks

File: python/README.md [hunk 10]
```diff
@@ -60,53 +60,55 @@ following 4 steps:
 ### Step 1: The user creates a registration request.
 
 ```python
-ctx, alpha = opaque.CreateRegistrationRequest(password)
+sec0, M = opaque.CreateRegistrationRequest(pwdU)
```
Comment (stef) [inline python/README.md:63]: secU is more expressive here

File: python/README.md [hunk 11]
```diff
@@ -60,53 +60,55 @@ following 4 steps:
...
+- `pwdU` is the user's password.
+
+The user should hold on to `sec0` securely until step 3 of the registration process.
+`M` needs to be passed to the server running step 2.
 
 ### Step 2: The server responds to the registration request.
 
 ```python
-sec, pub = opaque.CreateRegistrationResponse(alpha)
+sec1, pub = opaque.CreateRegistrationResponse(M)
```
Comment (stef) [inline python/README.md:74]: and here secS i guess is a better name.

File: python/opaque/__init__.py [hunk 12]
```diff
@@ -355,78 +355,89 @@ def UserAuth(ctx, auth, infos):
...
-#  is run by the user, taking as input the context ctx that was an
-#  output of the user running opaque_private_init_usr_start(), and the
-#  output pub from the server of opaque_private_init_srv_respond().
-#  The key parameter can be used as an extra contribution to the
-#  derivation of the rwd by means of being used as a key to the final
-#  hash, if not specified it uses the value specified by the rfc. The
+#  is run by the user, taking as input the context sec that was an
+#  output of the user running opaque_CreateRegistrationRequest(), and the
+#  output pub from the server of opaque_CreateRegistrationResponse().
+#  The info parameter can be used as an extra contribution to the
```
Comment (stef) [inline python/opaque/__init__.py:454]: is there still an info parameter?

Comment (creemama) [reply python/opaque/__init__.py]: It is not. I removed it in https://github.com/stef/libopaque/pull/13/commits/eef80c760d14ec90b6af0b9506216955ccd0ab42 .

File: python/test/simple.py (hunk 3 above)
Comment (creemama) [reply python/test/simple.py]: I added assertions.

File: python/test/simple.py [hunk 13]
```diff
@@ -74,4 +77,38 @@
...
+    ids=opaque.Ids("user", "server")
+    infos=opaque.App_Infos()
+    skS=pysodium.randombytes(crypto_scalarmult_SCALARBYTES);
+    # Uncomment the following if you compiled libopaque with -DNORANDOM -DTRACE and
+    # want the same output as register_with_global_server_key in
+    # src/tests/opaque-test.c. Also see a_randombytes in src/common.c.
+    #skS=ctypes.create_string_buffer(crypto_scalarmult_SCALARBYTES)
+    #for i in range(0, 32):
+    #    ctypes.memset(ctypes.addressof(skS) + i, i, 1)
+    pkS=pysodium.crypto_scalarmult_curve25519_base(skS);
```
Comment (creemama) [inline python/test/simple.py:97]: I was wondering. Why didn't `pysodium` have `crypto_scalarmult_base`?

Comment (stef) [reply python/test/simple.py:97]: i guess i never added it? and noone else bothered doing a PR?

**Comment by creemama** [comment]
We should probably rename `sec` in `opaque.h` to `secU` and `secS` where appropriate. I want to wait on that since some of the `sec*` variables might go away if we match the IETF CFRG more. The following is an example:
```
From the IETF CFRG draft:
skU, pkS, export_key = RecoverCredentials(pwdU, blind, response)

Our current API:
sk, authU, export_key, ids = opaque.RecoverCredentials(resp, secU, cfg, infos, pkS)
```

Looking at just the input arguments:
- `pwdU` is `secU.pwdU`.
- `blind` is `secU.blind`.
- `response.data` is `secU.M`.
- `response.pkS` is `pkS`.
- `response.envU` is `resp.envU`.

**Comment by stef** [comment]
> We should probably rename `sec` in `opaque.h` to `secU` and `secS` where appropriate. I want to wait on that since some of the `sec*` variables might go away if we match the IETF CFRG more. The following is an example:

i don't think the sec* are gonna go away. and to be honest i'm not hellbent on following the draft strictly, what matters to me is to be able to to have the interface between the client and the server be interoperable and be strict about that, so that it is possible to run the protocol with different implementations on both ends. naming, serialization (outside of the client/server boundary) and other things i don't care and i think it might even be contra productive.

> 
> ```
> From the IETF CFRG draft:
> skU, pkS, export_key = RecoverCredentials(pwdU, blind, response)

well yes, i think because the draft is a kind of "pure" opaque, which does not do the steps interleaved and leaves the concrete instantiation to an extra "chapter"

> Our current API:
> sk, authU, export_key, ids = opaque.RecoverCredentials(resp, secU, cfg, infos, pkS)
> ```
> 
> Looking at just the input arguments:
> 
>     * `pwdU` is `secU.pwdU`.
> 
>     * `blind` is `secU.blind`.
> 
>     * `response.data` is `secU.M`.
> 
>     * `response.pkS` is `pkS`.
> 
>     * `response.envU` is `resp.envU`.

they're all needed, they won't go away i'm quite sure.

**Pull request merged**